File: app/pfm/sketch_lines.py
"""Sketch Lines PFM — The foundational drawing algorithm for DrawingBotV3.

Algorithm (from documentation):
  1) Find the darkest area of the image
  2) Find the darkest pixel in that area
  3) Find the next darkest line from that pixel
  4) Brighten (erase) the part of the image that the line covers
  5) Repeat steps 2-4 until Squiggle Max Length is reached, then go to step 1
  Processing stops when Line Density or Line Max Limit is reached.
"""

from __future__ import annotations
import math
import numpy as np
from typing import List, Callable, Optional

from app.pfm import (
    PathFindingModule, PFMSetting, SettingType, DrawingGeometry,
    make_plotting_resolution_setting, make_random_seed_setting,
    make_sketch_style_settings, make_sketch_segment_settings,
    make_sketch_squiggle_settings, make_sketch_erase_settings,
    make_sketch_shading_settings,
)
from app.core.geometry import Path, Point


class SketchLinesPFM(PathFindingModule):
    """Transforms an image into lines using brightness data."""

    @property
    def name(self) -> str:
        return "Sketch Lines"

    @property
    def description(self) -> str:
        return "Transforms an image into lines using brightness data."

    @property
    def category(self) -> str:
        return "Sketch"

    @property
    def is_premium(self) -> bool:
        return False  # FREE

    def _define_settings(self) -> List[PFMSetting]:
        settings = [
            make_plotting_resolution_setting(),
            make_random_seed_setting(),
        ]
        settings.extend(make_sketch_style_settings())
        settings.extend(make_sketch_segment_settings())
        settings.extend(make_sketch_squiggle_settings())
        settings.extend(make_sketch_erase_settings())
        settings.extend(make_sketch_shading_settings())
        return settings

    def _process(self, image: np.ndarray,
                 progress: Callable) -> List[DrawingGeometry]:
        """Core Sketch Lines algorithm."""
        h, w = image.shape[:2]

        # Apply plotting resolution
        plot_res = self.get("plotting_resolution")
        if plot_res and plot_res != 1.0:
            new_w = max(1, int(w * plot_res))
            new_h = max(1, int(h * plot_res))
            import cv2
            work_img = cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_AREA)
        else:
            work_img = image.copy()
            new_w, new_h = w, h

        # Convert to float for precise brightness manipulation
        lightened = work_img.astype(np.float32)

        # Precompute edge maps if needed
        edge_map = None
        sobel_map = None
        direction_map = None

        edge_power = self.get("edge_power") or 0.0
        sobel_power = self.get("sobel_power") or 0.0
        directionality = self.get("directionality") or 0.0

        if edge_power > 0:
            import cv2
            edges = cv2.Canny(work_img, 50, 150)
            edge_map = edges.astype(np.float32) / 255.0

        if sobel_power > 0:
            import cv2
            sobelx = cv2.Sobel(work_img, cv2.CV_64F, 1, 0, ksize=3)
            sobely = cv2.Sobel(work_img, cv2.CV_64F, 0, 1, ksize=3)
            sobel_mag = np.sqrt(sobelx**2 + sobely**2)
            sobel_map = (sobel_mag / (sobel_mag.max() + 1e-10)).astype(np.float32)

        if directionality > 0:
            import cv2
            sobelx = cv2.Sobel(work_img, cv2.CV_64F, 1, 0, ksize=3)
            sobely = cv2.Sobel(work_img, cv2.CV_64F, 0, 1, ksize=3)
            direction_map = np.arctan2(sobely, sobelx).astype(np.float32)

        # Get settings
        line_density = self.get("line_density") or 75.0
        line_min_len = max(1, int(self.get("line_min_length") or 2))
        line_max_len = max(1, int(self.get("line_max_length") or 40))
        line_max_limit = int(self.get("line_max_limit") or -1)
        angle_tests = max(1, int(self.get("angle_tests") or 72))
        luminance_power = (self.get("luminance_power") or 100.0) / 100.0
        distortion = (self.get("distortion") or 0.0) / 100.0
        angularity_val = (self.get("angularity") or 0.0) / 100.0
        delta_angle = self.get("drawing_delta_angle") or 360.0

        sqg_min = max(0, int(self.get("squiggle_min_length") or 0))
        sqg_max = max(1, int(self.get("squiggle_max_length") or 500))
        sqg_deviation = (self.get("squiggle_max_deviation") or 25.0) / 100.0
        should_lift = self.get("should_lift_pen")
        if should_lift is None:
            should_lift = True

        erase_min = int(self.get("erase_min") or 50)
        erase_max = int(self.get("erase_max") or 125)
        erase_r_min = self.get("erase_radius_min") or 1.0
        erase_r_max = self.get("erase_radius_max") or 1.0
        tone = (self.get("tone") or 50.0) / 100.0

        # Shading settings
        shading_on = self.get("shading") or False
        shade_thresh = (self.get("shading_threshold") or 50.0) / 100.0
        shade_angle_min = math.radians(self.get("shading_start_angle_min") or -85.0)
        shade_angle_max = math.radians(self.get("shading_start_angle_max") or 95.0)
        shade_delta = math.radians(self.get("shading_delta_angle") or 360.0)

        # Calculate target brightness for density
        initial_brightness = float(lightened.mean())
        target_brightness = initial_brightness + (255.0 - initial_brightness) * (line_density / 100.0)

        # Precompute angle increments
        delta_rad = math.radians(abs(delta_angle)) if abs(delta_angle) < 360 else math.pi * 2
        angle_step = delta_rad / max(1, angle_tests)

        geometries: List[DrawingGeometry] = []
        total_segments = 0
        rng = self._rng

        # Scale factor for mapping back to original coords
        sx = w / new_w
        sy = h / new_h

        iteration = 0
        max_iterations = new_w * new_h  # Safety limit

        while iteration < max_iterations:
            if self.is_cancelled:
                break

            # Check stopping conditions
            current_brightness = float(lightened.mean())
            density_progress = (current_brightness - initial_brightness) / max(1.0, target_brightness - initial_brightness)
            density_progress = min(1.0, max(0.0, density_progress))

            if density_progress >= 1.0:
                break
            if line_max_limit > 0 and total_segments >= line_max_limit:
                break

            # Determine if we're in shading mode
            is_shading = shading_on and density_progress >= shade_thresh

            # Step 1: Find darkest area — sample random positions weighted by darkness
            # Create darkness map (invert lightened so dark = high value)
            darkness = 255.0 - lightened
            darkness = np.clip(darkness, 0, 255)

            # Weight by darkness
            darkness_sum = darkness.sum()
            if darkness_sum < 1.0:
                break  # Image fully brightened

            # Random weighted position selection
            flat = darkness.ravel()
            probs = flat / darkness_sum
            idx = rng.choices(range(len(flat)), weights=probs, k=1)[0]
            start_y = idx // new_w
            start_x = idx % new_w

            # Step 2-4: Build a squiggle from this start position
            squiggle: Path = [(float(start_x), float(start_y))]
            cx, cy = float(start_x), float(start_y)
            start_brightness_at_origin = lightened[start_y, start_x]
            segment_count = 0
            prev_angle = rng.uniform(0, 2 * math.pi)

            for seg_i in range(sqg_max):
                if self.is_cancelled:
                    break

                # Step 3: Find the darkest direction
                best_score = -float('inf')
                best_x, best_y = cx, cy
                best_angle = prev_angle

                # Determine base angle for shading
                if is_shading:
                    base_angle = rng.uniform(shade_angle_min, shade_angle_max)
                    test_delta = shade_delta
                else:
                    base_angle = prev_angle - delta_rad / 2
                    test_delta = delta_rad

                seg_len = rng.uniform(line_min_len, line_max_len)

                for t in range(angle_tests):
                    test_angle = base_angle + t * (test_delta / max(1, angle_tests))

                    # Add angularity (prefer continuing same direction)
                    if angularity_val > 0 and not is_shading:
                        angle_diff = abs(test_angle - prev_angle)
                        if angle_diff > math.pi:
                            angle_diff = 2 * math.pi - angle_diff
                        angularity_penalty = angle_diff / math.pi * angularity_val
                    else:
                        angularity_penalty = 0

                    # Add distortion
                    if distortion > 0:
                        test_angle += rng.gauss(0, distortion * 0.5)

                    # Calculate endpoint
                    ex = cx + math.cos(test_angle) * seg_len
                    ey = cy + math.sin(test_angle) * seg_len

                    # Bounds check
                    ex = max(0, min(new_w - 1, ex))
                    ey = max(0, min(new_h - 1, ey))

                    # Sample brightness along the line (simplified: use midpoint + endpoint)
                    mx, my = (cx + ex) / 2, (cy + ey) / 2
                    mx_i, my_i = int(mx), int(my)
                    ex_i, ey_i = int(ex), int(ey)

                    if 0 <= mx_i < new_w and 0 <= my_i < new_h and 0 <= ex_i < new_w and 0 <= ey_i < new_h:
                        # Score = darkness (lower brightness = better)
                        brightness_mid = lightened[my_i, mx_i]
                        brightness_end = lightened[ey_i, ex_i]
                        score = -(brightness_mid + brightness_end) / 2.0 * luminance_power

                        # Add edge detection bonus
                        if edge_map is not None and edge_power > 0:
                            edge_score = (edge_map[my_i, mx_i] + edge_map[ey_i, ex_i]) / 2.0
                            score += edge_score * edge_power

                        if sobel_map is not None and sobel_power > 0:
                            sobel_score = (sobel_map[my_i, mx_i] + sobel_map[ey_i, ex_i]) / 2.0
                            score += sobel_score * sobel_power

                        # Directionality bonus
                        if direction_map is not None and directionality > 0:
                            local_dir = direction_map[my_i, mx_i]
                            dir_diff = abs(test_angle - local_dir)
                            if dir_diff > math.pi:
                                dir_diff = 2 * math.pi - dir_diff
                            dir_score = (1.0 - dir_diff / math.pi)
                            score += dir_score * directionality * 0.5

                        score -= angularity_penalty * 50

                        if score > best_score:
                            best_score = score
                            best_x, best_y = ex, ey
                            best_angle = test_angle

                # Check squiggle deviation
                if seg_i > 0 and sqg_deviation < 1.0:
                    current_b = lightened[int(best_y), int(best_x)]
                    deviation = abs(current_b - start_brightness_at_origin) / 255.0
                    if deviation > sqg_deviation and segment_count >= sqg_min:
                        break

                # Add the segment
                squiggle.append((best_x, best_y))
                segment_count += 1
                total_segments += 1

                # Step 4: Erase (brighten) along the line
                self._erase_line(lightened, cx, cy, best_x, best_y,
                                 erase_min, erase_max, erase_r_min, erase_r_max,
                                 tone, new_w, new_h)

                cx, cy = best_x, best_y
                prev_angle = best_angle

                # Check min length for squiggle
                if segment_count >= sqg_max:
                    break

            # Scale squiggle back to original coordinates and add
            if len(squiggle) >= 2:
                scaled = [(px * sx, py * sy) for px, py in squiggle]
                geometries.append(DrawingGeometry(path=scaled, pen_index=0))

            # Progress update
            iteration += 1
            if iteration % 50 == 0:
                progress(density_progress, len(geometries),
                         f"Drawing... {len(geometries)} shapes, "
                         f"{density_progress*100:.0f}% density")

        return geometries

    def _erase_line(self, img: np.ndarray,
                    x0: float, y0: float, x1: float, y1: float,
                    erase_min: int, erase_max: int,
                    radius_min: float, radius_max: float,
                    tone: float,
                    w: int, h: int):
        """Brighten pixels along a line (the erasing process)."""
        dist = math.hypot(x1 - x0, y1 - y0)
        steps = max(1, int(dist))

        for s in range(steps + 1):
            t = s / max(1, steps)

            # Tone curve interpolation
            tone_t = t ** (1.0 / max(0.01, tone)) if tone > 0 else 0.5
            erase_val = erase_min + (erase_max - erase_min) * tone_t
            radius = radius_min + (radius_max - radius_min) * tone_t

            px = x0 + (x1 - x0) * t
            py = y0 + (y1 - y0) * t

            # Apply erasing in a circular area
            r_int = max(0, int(radius))
            px_i, py_i = int(px), int(py)

            if r_int == 0:
                # Single pixel
                if 0 <= px_i < w and 0 <= py_i < h:
                    img[py_i, px_i] = min(255.0, img[py_i, px_i] + erase_val)
            else:
                # Circular area
                y_lo = max(0, py_i - r_int)
                y_hi = min(h, py_i + r_int + 1)
                x_lo = max(0, px_i - r_int)
                x_hi = min(w, px_i + r_int + 1)

                for ey in range(y_lo, y_hi):
                    for ex in range(x_lo, x_hi):
                        d = math.hypot(ex - px, ey - py)
                        if d <= radius:
                            img[ey, ex] = min(255.0, img[ey, ex] + erase_val)

File: cpp_backup/pfm/sketch_lines.cpp
/**
 * sketch_lines.cpp — C++ port of the SketchLines PFM algorithm.
 *
 * Algorithm (from DrawingBotV3 docs):
 *   1) Find the darkest area of the image (weighted random sample)
 *   2) Find the darkest pixel in that area
 *   3) Find the next darkest line direction from that pixel (angle tests)
 *   4) Brighten (erase) the part of the image that the line covers
 *   5) Repeat steps 2-4 until squiggle max length is reached → step 1
 *   Stop when line density or line max limit is reached.
 */

#include "pfm/sketch_lines.h"
#include <cmath>
#include <opencv2/imgproc.hpp>

SketchLinesPFM::SketchLinesPFM(QObject* parent) : PathFindingModule(parent) {
    initSettings();
}

QVector<PFMSetting> SketchLinesPFM::defineSettings() const {
    return {
        // Default
        { "plotting_resolution", "Plotting Resolution", SettingType::Number,     1.0,  {}, 0.05,-1,-1,-1, 0.05, {}, "Default","Controls the resolution of the image." },
        { "random_seed",         "Random Seed",         SettingType::Integer,    42,   {}, 0, 999999, 0, 999999, 1, {}, "Default" },
        // Style
        { "should_lift_pen",     "Should Lift Pen",     SettingType::Boolean,    true, {}, 0,1,0,1,1,{}, "Style", "Lift pen between squiggles." },
        { "directionality",      "Directionality",      SettingType::Number,     0.0,  {}, 0, 200, 0, 100, 1, {}, "Style" },
        { "distortion",          "Distortion",          SettingType::Number,     0.0,  {}, 0, 200, 0, 100, 1, {}, "Style" },
        { "angularity",          "Angularity",          SettingType::Number,     0.0,  {}, 0, 200, 0, 100, 1, {}, "Style" },
        { "edge_power",          "Edge Power",          SettingType::Number,     0.0,  {}, 0, 200, 0, 100, 1, {}, "Style" },
        { "sobel_power",         "Sobel Power",         SettingType::Number,     0.0,  {}, 0, 200, 0, 100, 1, {}, "Style" },
        { "luminance_power",     "Luminance Power",     SettingType::Number,   100.0,  {}, 0, 200, 0, 200, 1, {}, "Style" },
        { "drawing_delta_angle", "Drawing Delta Angle", SettingType::Number,   360.0,  {}, -360,360,-360,360,1,{}, "Style" },
        // Segments
        { "line_density",        "Line Density",        SettingType::Percentage, 75.0, {}, 0,100,0,100,1,  {}, "Segments" },
        { "line_min_length",     "Line Min Length",     SettingType::Number,     2.0,  {}, 1,1000,2,500,1, {}, "Segments" },
        { "line_max_length",     "Line Max Length",     SettingType::Number,     40.0, {}, 1,1000,2,500,1, {}, "Segments" },
        { "line_max_limit",      "Line Max Limit",      SettingType::Integer,    -1,   {}, -1,1000000,-1,1000000,1,{}, "Segments" },
        { "angle_tests",         "Angle Tests",         SettingType::Integer,    72,   {}, 1,720,1,360,1,  {}, "Segments" },
        // Squiggles
        { "squiggle_min_length",    "Squiggle Min Length",    SettingType::Number, 0.0,  {}, 0,10000,0,5000,1, {}, "Squiggles" },
        { "squiggle_max_length",    "Squiggle Max Length",    SettingType::Number, 500.0,{}, 0,10000,0,5000,1, {}, "Squiggles" },
        { "squiggle_max_deviation", "Squiggle Max Deviation", SettingType::Percentage, 25.0, {}, 0,100,0,100,1, {}, "Squiggles" },
        // Erasing
        { "erase_min",           "Erase Min",           SettingType::Integer,    50,   {}, 0,255,0,255,1,  {}, "Erasing" },
        { "erase_max",           "Erase Max",           SettingType::Integer,    125,  {}, 0,255,0,255,1,  {}, "Erasing" },
        { "erase_radius_min",    "Erase Radius Min",    SettingType::Number,     1.0,  {}, 0,50,0,50,0.5,  {}, "Erasing" },
        { "erase_radius_max",    "Erase Radius Max",    SettingType::Number,     1.0,  {}, 0,50,0,50,0.5,  {}, "Erasing" },
        { "tone",                "Tone",                SettingType::Percentage,  50.0, {}, 0,100,0,100,1,  {}, "Erasing" },
        // Shading
        { "shading",             "Shading",             SettingType::Boolean,    false,{}, 0,1,0,1,1,{}, "Shading" },
        { "shading_threshold",   "Shading Threshold",   SettingType::Percentage, 50.0, {}, 0,100,0,100,1,{}, "Shading" },
        { "shading_start_angle_min", "Start Angle Min", SettingType::Number, -85.0, {}, -360,360,-360,360,1,{}, "Shading" },
        { "shading_start_angle_max", "Start Angle Max", SettingType::Number,  95.0, {}, -360,360,-360,360,1,{}, "Shading" },
        { "shading_delta_angle", "Shading Delta Angle", SettingType::Number,  360.0, {}, -360,360,-360,360,1,{}, "Shading" },
    };
}

// ---------------------------------------------------------------------------
// Erase (brighten) along a line — same math as Python _erase_line
// ---------------------------------------------------------------------------
void SketchLinesPFM::eraseLine(cv::Mat& img,
                                double x0, double y0, double x1, double y1,
                                double eraseMin, double eraseMax,
                                double radiusMin, double radiusMax,
                                double tone, int w, int h) {
    double dist = std::hypot(x1 - x0, y1 - y0);
    int steps = std::max(1, (int)dist);
    float* data = img.ptr<float>(0);

    for (int s = 0; s <= steps; ++s) {
        double t = (double)s / steps;
        double toneT = (tone > 0.0) ? std::pow(t, 1.0 / std::max(0.01, tone)) : 0.5;
        double eraseVal = eraseMin + (eraseMax - eraseMin) * toneT;
        double radius   = radiusMin + (radiusMax - radiusMin) * toneT;

        double px = x0 + (x1 - x0) * t;
        double py = y0 + (y1 - y0) * t;
        int pxi = (int)px, pyi = (int)py;

        int rInt = std::max(0, (int)radius);
        if (rInt == 0) {
            if (pxi >= 0 && pxi < w && pyi >= 0 && pyi < h)
                data[pyi * w + pxi] = std::min(255.0f, data[pyi * w + pxi] + (float)eraseVal);
        } else {
            int yLo = std::max(0, pyi - rInt), yHi = std::min(h, pyi + rInt + 1);
            int xLo = std::max(0, pxi - rInt), xHi = std::min(w, pxi + rInt + 1);
            for (int ey = yLo; ey < yHi; ++ey) {
                for (int ex = xLo; ex < xHi; ++ex) {
                    double d = std::hypot(ex - px, ey - py);
                    if (d <= radius)
                        data[ey * w + ex] = std::min(255.0f, data[ey * w + ex] + (float)eraseVal);
                }
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Main processing
// ---------------------------------------------------------------------------
QVector<DrawingGeometry> SketchLinesPFM::_process(const cv::Mat& image) {
    // ---- apply plotting resolution ----
    cv::Mat workImg;
    double plotRes = get("plotting_resolution").toDouble();
    if (plotRes > 0.0 && std::abs(plotRes - 1.0) > 1e-4) {
        int nw = std::max(1, (int)(image.cols * plotRes));
        int nh = std::max(1, (int)(image.rows * plotRes));
        cv::resize(image, workImg, cv::Size(nw, nh), 0, 0, cv::INTER_AREA);
    } else {
        workImg = image.clone();
    }
    const int W = workImg.cols, H = workImg.rows;

    // Working float buffer (mutable brightness)
    cv::Mat lightened;
    workImg.convertTo(lightened, CV_32F);

    // ---- Optional edge/sobel/direction maps ----
    const double edgePower   = get("edge_power").toDouble();
    const double sobelPower  = get("sobel_power").toDouble();
    const double directional = get("directionality").toDouble();

    cv::Mat edgeMap, sobelMap, directionMap;

    if (edgePower > 0) {
        cv::Mat edges;
        cv::Canny(workImg, edges, 50, 150);
        edges.convertTo(edgeMap, CV_32F, 1.0 / 255.0);
    }
    if (sobelPower > 0) {
        cv::Mat sx, sy, mag;
        cv::Sobel(workImg, sx, CV_64F, 1, 0, 3);
        cv::Sobel(workImg, sy, CV_64F, 0, 1, 3);
        cv::magnitude(sx, sy, mag);
        double mn, mx;
        cv::minMaxLoc(mag, &mn, &mx);
        mag.convertTo(sobelMap, CV_32F, 1.0 / (mx + 1e-10));
    }
    if (directional > 0) {
        cv::Mat sx, sy;
        cv::Sobel(workImg, sx, CV_64F, 1, 0, 3);
        cv::Sobel(workImg, sy, CV_64F, 0, 1, 3);
        // directionMap[y,x] = atan2(sy, sx)
        directionMap.create(H, W, CV_32F);
        for (int y = 0; y < H; ++y)
            for (int x = 0; x < W; ++x)
                directionMap.at<float>(y, x) = (float)std::atan2(
                    sy.at<double>(y, x), sx.at<double>(y, x));
    }

    // ---- Read settings ----
    const double lineDensity     = get("line_density").toDouble() / 100.0 * 100.0; // keep as %
    const int    lineMinLen      = std::max(1, (int)get("line_min_length").toDouble());
    const int    lineMaxLen      = std::max(1, (int)get("line_max_length").toDouble());
    const int    lineMaxLimit    = get("line_max_limit").toInt();
    const int    angleTests      = std::max(1, get("angle_tests").toInt());
    const double luminancePower  = get("luminance_power").toDouble() / 100.0;
    const double distortion      = get("distortion").toDouble() / 100.0;
    const double angularityVal   = get("angularity").toDouble() / 100.0;
    const double deltaAngle      = get("drawing_delta_angle").toDouble();

    const int    sqgMin          = std::max(0, (int)get("squiggle_min_length").toDouble());
    const int    sqgMax          = std::max(1, (int)get("squiggle_max_length").toDouble());
    const double sqgDeviation    = get("squiggle_max_deviation").toDouble() / 100.0;

    const double eraseMin        = get("erase_min").toDouble();
    const double eraseMax        = get("erase_max").toDouble();
    const double eraseRMin       = get("erase_radius_min").toDouble();
    const double eraseRMax       = get("erase_radius_max").toDouble();
    const double tone            = get("tone").toDouble() / 100.0;

    const bool   shadingOn       = get("shading").toBool();
    const double shadeThresh     = get("shading_threshold").toDouble() / 100.0;
    const double shadeAngleMin   = get("shading_start_angle_min").toDouble() * M_PI / 180.0;
    const double shadeAngleMax   = get("shading_start_angle_max").toDouble() * M_PI / 180.0;
    const double shadeDelta      = get("shading_delta_angle").toDouble() * M_PI / 180.0;

    // ---- Scale factors to map back to original coords ----
    const double sx = (double)image.cols / W;
    const double sy = (double)image.rows / H;

    // ---- Target brightness ----
    cv::Scalar meanScalar = cv::mean(lightened);
    const double initialBrightness = meanScalar[0];
    const double targetBrightness  = initialBrightness
        + (255.0 - initialBrightness) * (lineDensity / 100.0);

    // Pre-compute delta_rad
    const double deltaRad  = (std::abs(deltaAngle) < 360.0)
                             ? (std::abs(deltaAngle) * M_PI / 180.0)
                             : (2.0 * M_PI);

    float* lightenedPtr = lightened.ptr<float>(0);
    const float* edgePtr    = edgeMap.empty()    ? nullptr : edgeMap.ptr<float>(0);
    const float* sobelPtr   = sobelMap.empty()   ? nullptr : sobelMap.ptr<float>(0);
    const float* dirPtr     = directionMap.empty()? nullptr : directionMap.ptr<float>(0);

    QVector<DrawingGeometry> geometries;
    int totalSegments = 0;
    int iteration     = 0;
    const int maxIterations = W * H;

    // Darkness weights vector (reused)
    std::vector<double> weights(W * H);

    while (iteration < maxIterations && !isCancelled()) {
        // ---- Check stopping conditions ----
        double curBrightness = 0.0;
        for (int i = 0; i < W * H; ++i) curBrightness += lightenedPtr[i];
        curBrightness /= (W * H);

        double densityProgress = (curBrightness - initialBrightness)
                                 / std::max(1.0, targetBrightness - initialBrightness);
        densityProgress = std::clamp(densityProgress, 0.0, 1.0);

        if (densityProgress >= 1.0) break;
        if (lineMaxLimit > 0 && totalSegments >= lineMaxLimit) break;

        bool isShading = shadingOn && (densityProgress >= shadeThresh);

        // ---- Step 1: Weighted random start position ----
        double darknessSum = 0.0;
        for (int i = 0; i < W * H; ++i) {
            double d = std::max(0.0, 255.0 - (double)lightenedPtr[i]);
            weights[i] = d;
            darknessSum += d;
        }
        if (darknessSum < 1.0) break;

        int startIdx = weightedChoice(weights);
        int startY   = startIdx / W;
        int startX   = startIdx % W;

        // ---- Step 2-4: Build a squiggle ----
        Path squiggle;
        squiggle.push_back({ (double)startX, (double)startY });
        double cx0 = (double)startX, cy0 = (double)startY;
        double startBrightness = lightenedPtr[startY * W + startX];
        int segmentCount = 0;
        double prevAngle = randUniform(0.0, 2.0 * M_PI);

        for (int segI = 0; segI < sqgMax && !isCancelled(); ++segI) {
            // Step 3: find darkest direction
            double bestScore = -std::numeric_limits<double>::infinity();
            double bestX = cx0, bestY = cy0, bestAngle = prevAngle;

            double baseAngle, testDelta;
            if (isShading) {
                baseAngle = randUniform(shadeAngleMin, shadeAngleMax);
                testDelta = shadeDelta;
            } else {
                baseAngle = prevAngle - deltaRad / 2.0;
                testDelta = deltaRad;
            }

            double segLen = randUniform(lineMinLen, lineMaxLen);

            for (int t = 0; t < angleTests; ++t) {
                double testAngle = baseAngle + t * (testDelta / std::max(1, angleTests));

                // Angularity penalty
                double angularPenalty = 0.0;
                if (angularityVal > 0 && !isShading) {
                    double diff = std::abs(testAngle - prevAngle);
                    if (diff > M_PI) diff = 2.0 * M_PI - diff;
                    angularPenalty = diff / M_PI * angularityVal;
                }

                // Distortion
                if (distortion > 0)
                    testAngle += randGauss(0.0, distortion * 0.5);

                double ex = cx0 + std::cos(testAngle) * segLen;
                double ey = cy0 + std::sin(testAngle) * segLen;
                ex = std::clamp(ex, 0.0, (double)(W - 1));
                ey = std::clamp(ey, 0.0, (double)(H - 1));

                double mx = (cx0 + ex) / 2.0, my = (cy0 + ey) / 2.0;
                int mxi = (int)mx, myi = (int)my;
                int exi = (int)ex, eyi = (int)ey;

                if (mxi >= 0 && mxi < W && myi >= 0 && myi < H &&
                    exi >= 0 && exi < W && eyi >= 0 && eyi < H) {
                    double bMid = lightenedPtr[myi * W + mxi];
                    double bEnd = lightenedPtr[eyi * W + exi];
                    double score = -(bMid + bEnd) / 2.0 * luminancePower;

                    if (edgePtr) {
                        double edgeScore = ((double)edgePtr[myi*W+mxi] + edgePtr[eyi*W+exi]) / 2.0;
                        score += edgeScore * edgePower;
                    }
                    if (sobelPtr) {
                        double sobelScore = ((double)sobelPtr[myi*W+mxi] + sobelPtr[eyi*W+exi]) / 2.0;
                        score += sobelScore * sobelPower;
                    }
                    if (dirPtr && directional > 0) {
                        double localDir = dirPtr[myi * W + mxi];
                        double diff = std::abs(testAngle - localDir);
                        if (diff > M_PI) diff = 2.0 * M_PI - diff;
                        double dirScore = 1.0 - diff / M_PI;
                        score += dirScore * directional * 0.5;
                    }
                    score -= angularPenalty * 50.0;

                    if (score > bestScore) {
                        bestScore = score;
                        bestX = ex; bestY = ey;
                        bestAngle = testAngle;
                    }
                }
            }

            // Check squiggle deviation
            if (segI > 0 && sqgDeviation < 1.0) {
                int bxi = (int)bestX, byi = (int)bestY;
                if (bxi >= 0 && bxi < W && byi >= 0 && byi < H) {
                    double curB = lightenedPtr[byi * W + bxi];
                    double dev = std::abs(curB - startBrightness) / 255.0;
                    if (dev > sqgDeviation && segmentCount >= sqgMin)
                        break;
                }
            }

            squiggle.push_back({ bestX, bestY });
            ++segmentCount;
            ++totalSegments;

            eraseLine(lightened, cx0, cy0, bestX, bestY,
                      eraseMin, eraseMax, eraseRMin, eraseRMax, tone, W, H);

            cx0 = bestX; cy0 = bestY;
            prevAngle = bestAngle;
            if (segmentCount >= sqgMax) break;
        }

        // Scale back to original coords
        if (squiggle.size() >= 2) {
            Path scaled;
            scaled.reserve(squiggle.size());
            for (auto& p : squiggle)
                scaled.push_back({ p.first * sx, p.second * sy });
            geometries.append(DrawingGeometry{ scaled, 0 });
        }

        ++iteration;
        if (iteration % 50 == 0) {
            emitProgress((float)densityProgress, geometries.size(),
                         QString("Drawing... %1 shapes, %2% density")
                         .arg(geometries.size())
                         .arg((int)(densityProgress * 100)));
        }
    }
    return geometries;
}
